inject/src/main/java/io/micronaut/context/AnnotationReflectionUtils.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.context;
+
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.type.Argument;
+import io.micronaut.inject.annotation.AnnotationMetadataHierarchy;
+import io.micronaut.inject.annotation.MutableAnnotationMetadata;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.AnnotatedParameterizedType;
+import java.lang.reflect.AnnotatedType;
+import java.lang.reflect.GenericArrayType;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.lang.reflect.WildcardType;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The annotation reflection utils.
+ *
+ * @author Denis Stepanov
+ * @since 4.6.0
+ */
+@Internal
+public final class AnnotationReflectionUtils {
+
+    AnnotationReflectionUtils() {}
+
+    /**
+     * Find implementation as an argument with types and annotations.
+     *
+     * @param implementationType The implementation class of the interface
+     * @param implementedType    The implementedType type - interface or an abstract class
+     * @return The argument of the interface with types and annotations
+     * @since 4.6
+     */
+    @Nullable
+    public static Argument<?> findImplementationAsArgument(@NonNull Class<?> implementationType,
```

Comment:
can we rename this `resolveArgumentFromGenerics` the name here is confusing imo

Reply:
Is it understandable that it will return `Argument.of(implementedType, ...)` not just a generic type?